### DE analysis of Inflammation and Patient dataset ----
# counting matrix generated by TEtranscript
data <- read.table("/home/rpg18/Desktop/TFM/Server_alice2/Pipelines_scripts/TEtranscript/Sample_infl/rev_infl_changed.cntTable",header=T,row.names=1)
# meta-data csv file that stores SampleID and tissue type (four types) information of the samples
tissue <-read.csv("/home/rpg18/Desktop/TFM/Server_alice2/Pipelines_scripts/TEtranscript/Sample_infl/tissue_type_new.csv", row.names = "SampleID")
# meta-data csv file that stores SampleID and method type (Dmso or Rna Later) information of the samples
dmso <- read.csv("/home/rpg18/Desktop/TFM/Server_alice2/Pipelines_scripts/TEtranscript/Sample_infl/dmso_RNAlater_new.csv", row.names = "SampleID")
# meta-data csv file that stores SampleID and age information of the samples
age <- read.csv("/home/rpg18/Desktop/TFM/Server_alice2/Pipelines_scripts/TEtranscript/Sample_infl/age_new.csv", row.names = "SampleID")
#age$age <- as.character(age$age)

# creates groups variables with both Inflamed and Control condition levels
groups <- factor(c(rep("Inflamed",35),rep("Control",32)))

# minimum count per gene: 1
min_read <- 1
# removes rows/genes where the number of reads is less than min_read
data <- data[apply(data,1,function(x){max(x)}) > min_read,]
# generates a dataframe with SampleID as row names and a meta-data column (groups) with condition information across samples (Inflamed or Control)
sampleInfo <- data.frame(groups, row.names=colnames(data))
# merges tissue column to sampleInfo dataframe (dataframe: SampleID, groups, tissue)
meta_tissue <- merge(sampleInfo, tissue["tissue"],by="row.names", sort = FALSE, keep_order = 1)
# convert meta_tissue to dataframe
meta_tissue <- data.frame(meta_tissue, row.names = "Row.names")
# merges age column to meta_tissue dataframe (dataframe: SampleID, groups, tissue, age)
meta_age <- merge(meta_tissue, age["age"],by="row.names", sort = FALSE, keep_order = 1)
# convert meta_tissue to dataframe
meta_age <- data.frame(meta_age, row.names = "Row.names")

# merges method column (Dmso or Rna Later) to meta_age dataframe (dataframe: SampleID, groups, tissue, age, method)
meta_ALLdata <- merge(meta_age, dmso["method"],by="row.names", sort = FALSE, keep_order = 1)
# convert meta_ALLdata to dataframe
meta_ALLdata <- data.frame(meta_ALLdata, row.names = "Row.names")
meta_ALLdata


## Libraries: ----
library(DESeq2, quietly=T)
library(ggplot2)
library(genefilter)
library(pheatmap)
library(apeglm)
library(ReportingTools)
library(dplyr)
library(genefilter)
library(RColorBrewer)
library(PoiClaClu)
library(pcaExplorer)
library(tibble)
library(scales)



### dds_infl Object and DESeq2 analysis for Inflammation dataset: RnaLater vs Dmso + Inflamed vs Control ----
# DESeq object: dds
dds_infl <- DESeqDataSetFromMatrix(countData = data, colData = meta_ALLdata, design = ~ method + groups) #this design order if I don't specified in results(), log2 fold change and Wald test pvalue will be for the last variable (groups)
data
## relevel with Inflamed vs Control condition
dds_infl$groups = relevel(dds_infl$groups, ref = "Control") # levels specified by ref appears first
# DE analysis performed by DESeq()
dds_infl <- DESeq(dds_infl)
resultsNames(dds_infl)

## Now, one can relevel (re-order) the analysis with the condition of interest: "Rna Later" or "Control" to study the method or inflammation effects, respectively
#dds_infl$method = relevel(dds_infl$method, ref = "RNA_Later") # levels specified by ref appears first
# DE analysis performed by DESeq()
#dds_infl <- DESeq(dds_infl)
#resultsNames(dds_infl)

# extracts a result table from a DESeq analysis. Result information regarding base means across samples, log2FoldChanges, standard errors, test statistics, p-values and adjusted p-values (padj)
res.gr <- results(dds_infl, contrast=list("groups_Inflamed_vs_Control"),independentFiltering=F) # no eliminate genes from the analysis
#res.gr <- results(dds_infl, contrast=c("groups","Inflamed","Control"),independentFiltering=F)
res.met
# extracts a result table from a DESeq analysis. Result information regarding base means across samples, log2FoldChanges, standard errors, test statistics, p-values and adjusted p-values (padj)
res.met <- results(dds_infl, contrast=list("method_RNA_Later_vs_Dmso"),independentFiltering=F) # no eliminate genes from the analysis
#res.met <- results(dds_met, contrast=c("method","Dmso","RNA_Later"),independentFiltering=F)

### dds_pat Object and DESeq2 analysis for Patients dataset: Control vs Tumour ----
# counting matrix generated by TEtranscript
data_pat <- read.table("/home/rpg18/Desktop/TFM/Pipelines/tetoolkit/Sample_pat_TE/all_TE_changed.cntTable",header=T,row.names=1)

# creates group variables with both Tumour and Control condition levels
gr_pat <- factor(c(rep("Tumour",5),rep("Control",5)))

# minimum count per gene: 1
min_read <- 1
# removes rows/genes where the number of reads is less than min_read
data_pat <- data_pat[apply(data_pat,1,function(x){max(x)}) > min_read,]
# generates a dataframe with SampleID as row names and a meta-data column (gr_pat) with condition information across samples (Tumour or Control)
sampleInfo_pat <- data.frame(gr_pat,row.names=colnames(data_pat))
# add new column to sampleInfo_pat dataframe with name samples (for further analysis of differences between individual samples)
sampleInfo_pat$name <- colnames(data_pat) 
sampleInfo_pat

# DESeq object: dds_pat
dds_pat <- DESeqDataSetFromMatrix(countData = data_pat, colData = sampleInfo_pat, design = ~ gr_pat)
dds_pat$gr_pat= relevel(dds_pat$gr_pat,ref = "Control")
dds_pat  <- DESeq(dds_pat)

# extracts a result table from a DESeq analysis. Result information regarding base means across samples, log2FoldChanges, standard errors, test statistics, p-values and adjusted p-values (padj)
#res.pat <- results(dds_pat,independentFiltering=F) # independentFiltering was not necessary in this case
resultsNames(dds_pat)
res.pat <- results(dds_pat, contrast=list("gr_pat_Tumour_vs_Control"),independentFiltering=F)
res.pat
# filtering for significant DEG
resSig.pat <- res.pat[(!is.na(res.pat$padj) & (res.pat$padj < 0.0500) & (abs(res.pat$log2FoldChange) > 0.00000)),]
resSig.pat


### dds_inflsal: SalmonTE ----
data_inflsal <- round(dat$count)
data_inflsal
# creates group variables with both Tumour and Control condition levels
tissue_sal <-read.csv("/home/rpg18/Desktop/TFM/Pipelines/SalmonTE/out_all/tissue_type_new.csv", row.names = "SampleID")
# meta-data csv file that stores SampleID and method type (Dmso or Rna Later) information of the samples
dmso_sal <- read.csv("/home/rpg18/Desktop/TFM/Pipelines/SalmonTE/out_all/dmso_RNAlater_new.csv", row.names = "SampleID")
# meta-data csv file that stores SampleID and age information of the samples
age_sal <- read.csv("/home/rpg18/Desktop/TFM/Pipelines/SalmonTE/out_all/age_new.csv", row.names = "SampleID")
# condition file (inflamed vs control)
group_inflsal <- read.csv("/home/rpg18/Desktop/TFM/Pipelines/SalmonTE/out_all_2/condition.csv", row.names = "SampleID")
# replace treatment condition by inflamed
group_inflsal <- group_inflsal %>%
  mutate(condition = as.character(condition)) %>% 
  mutate(condition = replace(condition, condition == 'treatment', 'inflamed')) %>% 
  mutate(condition = as.factor(condition))

# generates a dataframe with SampleID as row names and a meta-data column (groups) with condition information across samples (Inflamed or Control)
sampleInfo_inflsal <- data.frame(group_inflsal, row.names=colnames(data_inflsal))
# merges tissue column to sampleInfo dataframe (dataframe: SampleID, groups, tissue)
meta_tissue_inflsal <- merge(sampleInfo_inflsal, tissue_sal["tissue"],by="row.names", sort = FALSE, keep_order = 1)
# convert meta_tissue to dataframe
meta_tissue_inflsal <- data.frame(meta_tissue_inflsal, row.names = "Row.names")
# merges age column to meta_tissue dataframe (dataframe: SampleID, groups, tissue, age)
meta_age_inflsal <- merge(meta_tissue_inflsal, age_sal["age"],by="row.names", sort = FALSE, keep_order = 1)
# convert meta_tissue to dataframe
meta_age_inflsal <- data.frame(meta_age_inflsal, row.names = "Row.names")

# merges method column (Dmso or Rna Later) to meta_age dataframe (dataframe: SampleID, groups, tissue, age, method)
meta_ALLdata_inflsal <- merge(meta_age_inflsal, dmso_sal["method"],by="row.names", sort = FALSE, keep_order = 1)
# convert meta_ALLdata to dataframe
meta_ALLdata_inflsal <- data.frame(meta_ALLdata_inflsal, row.names = "Row.names")
#col_data <- dat$col_data
# DESeq object: dds
dds_inflsal <- DESeqDataSetFromMatrix(countData = data_inflsal, colData = meta_ALLdata_inflsal, design = ~ method + condition) #this design order if I don't specified in results(), log2 fold change and Wald test pvalue will be for the last variable (groups)

## Now, one can relevel (re-order) the analysis with the condition of interest: "Rna Later" or "Control" to study the method or inflammation effects, respectively
dds_inflsal$condition = relevel(dds_inflsal$condition, ref = "control") # levels specified by ref appears first
# filtering of counts in dds object
dds_inflsal <- dds_inflsal[ rowSums(counts(dds_inflsal)) > 10, ]
# DE analysis performed by DESeq()
dds_inflsal <- DESeq(dds_inflsal)
resultsNames(dds_inflsal)  # "method_RNA_Later_vs_Dmso" and "condition_inflamed_vs_control"

# extracts a result table from a DESeq analysis. Result information regarding base means across samples, log2FoldChanges, standard errors, test statistics, p-values and adjusted p-values (padj)
res.grsal <- results(dds_inflsal, contrast=list("condition_inflamed_vs_control"),independentFiltering=F) # no eliminate genes from the analysis
# results for methods (dmso vs RNA later)
res.metsal <- results(dds_inflsal, contrast=list("method_RNA_Later_vs_Dmso"),independentFiltering=F)
# check results
res.grsal[which.min(res.grsal$padj),] # MER9B
results_inflsal <- read.csv("/home/rpg18/Desktop/TFM/Pipelines/SalmonTE/out_all_2/test/results.csv", row.names = "name")
results_inflsal[which.min(results_inflsal$padj),]


### dds_patsal: SalmonTE ----
data_patsal <- round(dat$count)
#data_pat <- read.table("/home/rpg18/Desktop/TFM/Pipelines/tetoolkit/Sample_pat_TE/all_TE_changed.cntTable",header=T,row.names=1)
data_patsal
# creates group variables with both Tumour and Control condition levels
#gr_patsal <- read.csv("/home/rpg18/Desktop/TFM/Pipelines/SalmonTE/Salmon_Patients_fastq4/condition.csv", row.names = "SampleID")
gr_patsal <- dat$col_data
# generates a dataframe with SampleID as row names and a meta-data column (groups) with condition information across samples (Inflamed or Control)
sampleInfo_patsal  <- data.frame(gr_patsal, row.names=colnames(data_patsal))
sampleInfo_patsal$name <- colnames(data_patsal) 

# DESeq object: dds_pat
dds_patsal <- DESeqDataSetFromMatrix(countData = data_patsal, colData = sampleInfo_patsal, design = ~ condition)
## relevel with Inflamed vs Control condition
dds_patsal$condition = relevel(dds_patsal$condition,ref = "control")
# filtering of counts in dds object
dds_patsal <- dds_patsal[ rowSums(counts(dds_patsal)) > 10, ]
# DE analysis performed by DESeq()
dds_patsal <- DESeq(dds_patsal)

# extracts a result table from a DESeq analysis. Result information regarding base means across samples, log2FoldChanges, standard errors, test statistics, p-values and adjusted p-values (padj)
res.patsal <- results(dds_patsal,independentFiltering=F) # independentFiltering was not necessary in this case
resultsNames(dds_patsal)
res.patsal <- results(dds_patsal, contrast=list("condition_tumour_vs_control"),independentFiltering=F)
res.patsal
res.patsal[which.min(res.patsal$padj),] # LTR12B
results_patsal <- read.csv("/home/rpg18/Desktop/TFM/Pipelines/SalmonTE/Salmon_out4_0.05/results.csv", row.names = "name")
results_patsal[which.min(results_patsal$padj),] # LTR12B
# filtering for significant DEG
resSig.patsal <- res.patsal[(!is.na(res.patsal$padj) & (res.patsal$padj < 0.0500) & (abs(res.patsal$log2FoldChange) > 0.00000)),]
resSig.patsal


### Write result tables ----
path_res <- '/home/rpg18/Desktop/TFM/Pipelines/SCRIPTS/Bedtools_trial/TEtrans_out/results/'
restable(res.gr,"infl.txt_0.4",path_res,res.gr$log2FoldChange,0.4)      # Inflammation - TEtranscripts
restable(res.grsal,"infl_TE_0.4",path_res,res.grsal$log2FoldChange,0.4)    # Inflamamation-SalmonTE
restable(res.met,"met_DEG",path_res,-1.0)     # method-TEtranscripts (Dmso.vs.RNA_later, in inflammation)
restable(res.metsal,"met_TE",path_res,-1.0)   # method-SalmonTE (Dmso.vs.RNA_later, in inflammation)
restable(res.pat,"pat.txt_0.4",path_res,res.pat$log2FoldChange,0.4)      # Patient-TETranscripts 
restable(res.patsal,"pat_TE_0.4",path_res,res.patsal$log2FoldChange,0.4)    # Patient-SalmonTE

restable <- function(res,name,path,cond1,value1){
  pathset <- setwd(path)  # set path
  # filtering for significant DEG
  resSig <- res[(!is.na(res$padj) & (res$padj < 1.00000) & (abs(cond1) > value1)), ]   # DE gene or TE
  # orders by log2FoldChange value
  resOrdered <- resSig[order(resSig$padj),]
  # writes output tables
  write.table(resOrdered, file=name,sep="\t", quote=F) # ordered by log2FoldChange
}



### rlog (or vst) for plotPCA function YES----
plotPCA_samples(dds_infl,"groups","tissue")
plotPCA_samples(dds_infl,"groups","method")
plotPCA_samples(dds_pat,"gr_pat","name")
plotPCA_samples(dds_inflsal,"condition","condition")
plotPCA_samples(dds_inflsal,"condition","method")
plotPCA_samples(dds_patsal,"condition","condition")
dds_infl$groups
# used
plotPCA_samples <- function(dds,cond1,cond2) {
  rld <- varianceStabilizingTransformation(dds)
  plotPCA(rld, intgroup = c(cond1,cond2))
}


### Table topGenes and expression Plotting of TOP down and up-regulated genes for both conditions: inflamed and tumour ----
path <- setwd("/home/rpg18/Desktop/TFM/Pipelines/SCRIPTS/Bedtools_trial/TEtrans_out/")
file_up <- "upReg"
file_down <- "downReg"

# up-regulated
upreg <- function(dds,res,cond,filename,pathname,top) {
  resUpReg = res[which(res$log2FoldChange > 0), ] # get the up-regulated genes
  top5_upGenes <- head(resUpReg[order(resUpReg$padj),], top) # order by padjusted and print the top 5
  #write.table(top5_upGenes, file.path(path,paste(cond,filename,sep="",collapse=NULL)), quote=F)
  rownames(top5_upGenes) # print top5 up-regulated genes
}
# down-regulated
downreg <- function(dds,res,cond,filename,pathname,top) {
  resDownReg = res[which(res$log2FoldChange < 0), ] # get the down-regulated genes
  top5_downGenes <- head(resDownReg[order(resDownReg$padj),], top) # order by padjusted 
  #write.table(top5_downGenes, file.path(path,paste(cond,filename,sep="",collapse=NULL)), quote=F)
  rownames(top5_downGenes) # print top5 down-regulated genes
}


### Ploting counts vs conditions YES: BOXPLOT coloured----
# code-reference: https://www.michaelchimenti.com/2018/12/publication-quality-boxplot-deseq2-data/ http://www.sthda.com/english/wiki/ggplot2-box-plot-quick-start-guide-r-software-and-data-visualization

## TEtranscripts
top_up_infl <- upreg(dds_infl,res.gr,"infl_",file_up,path,1)           # SLC6A14 for Inflamed condition
top_down_infl <-downreg(dds_infl,res.gr,"infl_",file_down,path,1)      # CNTN4 for Inflamed condition
top_up_met <- upreg(dds_infl,res.met,"inf_RNALater",file_up,path,1)    # FOLR2 for method condition (up-RNALater)
top_down_met <- downreg(dds_infl,res.met,"inf_Dmso_",file_down,path,1) # SF1 for method condition (up-Dmso)
top_up_pat <- upreg(dds_pat,res.pat,"pat_",file_up,path,1)             # CEMIP for patients dataset 
top_down_pat <- downreg(dds_pat,res.pat,"pat_",file_down,path,1)       # BEST4 for patients dataset

## SalmonTE
top_up_inflsal <- upreg(dds_inflsal,res.grsal,"inflsal_",file_up,path,1)        # MER66_I up in inflamed https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-018-4901-9 
top_down_inflsal <- downreg(dds_inflsal,res.grsal,"inflsal_",file_down,path,1)  # MER9B down in inflamed
top_up_patsal <- upreg(dds_patsal,res.patsal,"patsal_",file_down,path,1)        # MER57F up in tumour
top_down_patsal <- downreg(dds_patsal,res.patsal,"patsal_",file_up,path,1)      # LTR12B down in tumour
top_up_metsal <- upreg(dds_inflsal,res.metsal,"inf_RNALater",file_up,path,1)    # LTR22B2 up-RNALater (for method condition)
top_down_metsal <- downreg(dds_inflsal,res.metsal,"inf_Dmso_",file_down,path,1) # SVA_F up-Dmso (for method condition)

genename <- top_up_patsal
genename <- "MER66_I"

## SalmonTE
# inflammation dataset
boxes(dds_inflsal,dds_inflsal$condition,dds_inflsal$method,dds_inflsal$age,genename,res.grsal)
boxes_2(dds_inflsal,dds_inflsal$condition,dds_inflsal$method,genename,res.metsal)
boxes_basic(dds_inflsal,dds_inflsal$condition,dds_inflsal$condition,genename,res.grsal,"purple")
#boxes_basic(dds_inflsal,dds_inflsal$method,dds_inflsal$method,genename,res.metsal)
# patient dataset
boxes_basic(dds_patsal,dds_patsal$condition,dds_patsal$condition,genename,res.patsal,"burlywood")

genename <- "L1M3a:L1:LINE"
## TEtranscripts
# inflammation dataset
boxes(dds_infl,dds_infl$groups,dds_infl$method,dds_infl$age,genename,res.gr)
boxes_2(dds_infl,dds_infl$groups,dds_infl$method,genename,res.met)
boxes_basic(dds_infl,dds_infl$groups,dds_infl$groups,genename,res.gr,"purple")
#boxes_basic(dds_infl,dds_infl$method,dds_infl$method,genename,res.met)
# patient dataset
boxes_basic(dds_pat,dds_pat$gr_pat,dds_pat$gr_pat,genename,res.pat,"burlywood")

## Plotting functions
boxes <- function(dds,uno,dos,tres,gene,res) {
  traf1_counts <- log2(counts(dds[gene,], normalized = TRUE)+1)
  #m <- list(counts = as.numeric(traf1_counts), group = factor(uno,c("Inflamed","Control")), method = factor(dos,c("Dmso","RNA_Later")))#, method = as.factor(dds_infl$method))
  m <- list(counts = as.numeric(traf1_counts), group = as.factor(uno), method = as.factor(dos), age = as.factor(tres))#, method = as.factor(dds_infl$method))
  m <- as_tibble(m)
  # simple and perfect
  ggplot(m, aes(x = method, y = counts, color = group)) +
    geom_boxplot(#alpha=0.5,notch=F,
      #notchwidth = 0.5,
      outlier.colour='red',
      outlier.fill='red',
      outlier.size=1) +
    #geom_jitter(shape=10, position=position_jitter(0.2), color = method)+
    #scale_color_brewer(palette="Dark2")+
    #scale_color_manual(values=c("#E69F00", "#56B4E9"))+
    facet_grid(. ~ age)+
    scale_color_manual(values=c("#56B4E9","purple"))+
    #scale_y_continuous(trans = log2_trans())+
    #scale_y_continuous(trans = log2_trans(),
    #                  breaks = trans_breaks("log2", function(x) 2^x),
    #                 labels = trans_format("log2", math_format(2^.x)))+
    #scale_color_grey()+
    theme(
      #legend.position="none",
      plot.title = element_text(size=11)
    ) +
    ggtitle(paste(gene,"padj = ",res[gene,]$padj)) +
    xlab("")  + ylab("log2-Expression of normalised counts")
}
boxes_2 <- function(dds,uno,dos,gene,res) {
  traf1_counts <- log2(counts(dds[gene,], normalized = TRUE)+1)
  m <- list(counts = as.numeric(traf1_counts), group = as.factor(uno), method = as.factor(dos))#, method = as.factor(dds_infl$method))
  m <- as_tibble(m)
  # simple and perfect
  ggplot(m, aes(x = group, y = counts, color = method)) +
    geom_boxplot(#alpha=0.5,notch=F,
      #notchwidth = 0.5,
      outlier.colour='red',
      outlier.fill='red',
      outlier.size=1) +
    # geom_jitter(shape=10, position=position_jitter(0.2), color = method)+
    scale_color_brewer(palette="Dark2")+
    scale_fill_manual(values=c("grey","cyan","#E69F00", "#56B4E9"))+
    #facet_grid(. ~ age)+
    #scale_y_continuous(trans = log2_trans())+
    #scale_y_continuous(trans = log2_trans(),
    #                  breaks = trans_breaks("log2", function(x) 2^x),
    #                 labels = trans_format("log2", math_format(2^.x)))+
    #scale_color_grey()+
    theme(
      #legend.position="none",
      plot.title = element_text(size=11)
    ) +
    ggtitle(paste(gene,"padj = ",res[gene,]$padj)) +
    xlab("")  + ylab("log2-Expression counts")
}
boxes_basic <- function(dds,uno,dos,gene,res,colour) {
  traf1_counts <- log2(counts(dds[gene,], normalized = TRUE)+1)
  m <- list(counts = as.numeric(traf1_counts), group = as.factor(uno), method = as.factor(dos))#, method = as.factor(dds_infl$method))
  m <- as_tibble(m)
  # simple and perfect
  ggplot(m, aes(x = group, y = counts, color = group)) +
    geom_boxplot(#alpha=0.5,notch=F,
      #notchwidth = 0.5,
      outlier.colour='red',
      outlier.fill='red',
      outlier.size=1) +
    geom_jitter(shape=10, position=position_jitter(0.2))+
    scale_color_manual(values=c("#56B4E9",colour))+
    theme(
      #legend.position="none",
      plot.title = element_text(size=11)
    ) +
    ggtitle(paste(gene,"padj = ",res[gene,]$padj)) +
    xlab("")  + ylab("log2-Expression counts")
}

pheatmap(samplePoisDistMatrix,
         clustering_distance_rows=poisd$dd,
         clustering_distance_cols=poisd$dd,
         col=colors)


### Heatmap top DEGenes_rld YES----
topheatmap(dds_pat,100,"gr_pat","name")
topheatmap(dds_infl,50,"groups","method")
topheatmap(dds_patsal,100,"condition","condition")
topheatmap(dds_inflsal,50,"condition","tissue")

topheatmap <- function(dds,topn,cond1,cond2) {
  rld <- varianceStabilizingTransformation(dds)
  topVarGenes <- head(order(-rowVars(assay(rld))),topn) # first 50
  mat <- assay(rld)[ topVarGenes, ]
  mat <- mat - rowMeans(mat)
  df <- as.data.frame(colData(rld)[,c(cond1,cond2)])
  pheatmap(mat, annotation_col=df)
}
