
#!/bin/bash
####(All scripts used by run_ALL_comparison.sh need to be stored in the same directory)
### This script creates the output path seleted by the user, where multiple annotation bed files will be stored. Run it as: sh run_ALL_comparison.sh <output_folder> 
''' Script that generates the @/finalgenes_compared_func.txt and @/final_output.txt files '''
echo \n
## Path requirements
out_annot=$@
var2='./Annotation_hg19/geneID_UCSC.txt' # this file will be generated by lct2geneID.py script (a file storing just RefSeq accession numbers for all shared-LCTs)

## Pipeline starts:
# filters TECDetec output gtf files and generated filtered bed files:
sh run_filtering.sh
# generates multiple files stored in ./Out_bedtools folders and in $out_annot, such as sorted_only_inflamed_LCT.bed and sorted_only_tum_LCT.bed:
sh run_comparison_AnnotLCT.sh $out_annot		
# converts RefSeq-accessions to gene names stored in ./Annotation_hg19/annotated_genes.txt, and creates $@/full_info.txt that adds gene name to $@/annot_sorted.bed:
python3 lct2geneID.py $out_annot $var2	# it takes $out_annot/annot_sorted.bed and ./Annotation_hg19/tablesUCSC.txt (https://genome.ucsc.edu/cgi-bin/hgTables) files, and generates $var2 (file with RefSeq-accession) and ./Annotation_hg19/annotated_genes.txt files
			
## Now, one needs to extract significant annotated genes related to LCTs in both inflammation and tumour conditions. This is done by comparing TEtranscripts output with TECDetec output: Compare list of genes that are DE in both inflammation and tumour with the ./Annotation_hg19/annotated_genes.txt file that stores annotated genes related to shared-LCTs:
sh run_comparison_LCT_DEgenes.sh $out_annot	# it takes ./Annotation_hg19/annotated_genes.txt and both DEG text files for each condition (patient and inflammation), and gives $@/finalgenes_compared_func.txt (list of gene names) and $@/final_output.txt (complete list with LINE-1-5'end annotation information, chr strand and position, and RefSeq-accession and gene name for the specific LCT)

## Remove intermediary files (optional)
#rm -r 

## Find new output files in:
# $@/
# ./Bedtools_Out
# ./Annotation_hg19/
